tests/test_version_utils.py

```diff
@@ -0,0 +1,293 @@
+    assert spec_1 <= spec_1
+    assert spec_1 >= spec_1
+
+
+def test_version_spec_ge():
+    """Test VersionSpec()."""
+    spec_2 = VersionSpec.from_str('>=1.0.0')
+    assert not spec_2.contains('0.0.1')
+    assert spec_2.contains('1.0.0')
+    assert spec_2.contains('2')
+
+
+def test_version_spec_eq():
+    """Test VersionSpec()."""
+    spec_3 = VersionSpec.from_str('==1.0.0')
+    assert spec_3.contains('1.0.0')
+    assert spec_3.contains('1')
+    assert not spec_3.contains('2')
+    assert not spec_3.contains('0')
+
+    assert spec_3 == spec_3
+    assert not spec_3 != spec_3
+    assert not spec_3 < spec_3
+    assert not spec_3 > spec_3
+    assert spec_3 <= spec_3
+    assert spec_3 >= spec_3
+
+
+def test_version_spec():
+    """Test VersionSpec()."""
+    spec_1 = VersionSpec.from_str('<=2.1.1')
+    spec_2 = VersionSpec.from_str('>=1.0.0')
+    spec_3 = VersionSpec.from_str('==1.0.0')
+
+    assert not spec_1 == spec_2
+    assert spec_1 != spec_2
+    assert not spec_1 < spec_2
+    assert spec_1 > spec_2
+    assert not spec_1 <= spec_2
+    assert spec_1 >= spec_2
+
+    assert str(spec_1) == '<=2.1.1'
+    assert str(spec_2) == '>=1.0.0'
+    assert str(spec_3) == '==1.0.0'
+
+
+def test_version_range():
+    """Test VersionRange()."""
+    ver_range = VersionRange('<=6.5.4', '>=5.0.0')
+    assert ver_range.contains('6.0.0')
+    assert not ver_range.contains('4.0.0')
+    assert str(ver_range) == '<=6.5.4,5.0.0'
+
+    ver_range = VersionRange('>=5.0.0', '<=6.5.4')
+    assert ver_range.contains('6.0.0')
+    assert not ver_range.contains('4.0.0')
+    assert str(ver_range) == '<=6.5.4,5.0.0'
+
+    ver_range = VersionRange('==1.0.0')
+    assert ver_range.contains('1.0.0')
+    assert not ver_range.contains('0.0.1')
+    assert str(ver_range) == '==1.0.0'
+
+
+def test_version_range_from_list():
+    """Test VersionRange.from_list()."""
+    ver_range = VersionRange.from_list(['1', '2', '3', '4'], is_right_closed=True)
+    assert str(ver_range) == '<=4'
+
+    ver_range = VersionRange.from_list(
+        ['1', '2', '3', '4'], is_right_closed=True, is_left_closed=True
+    )
+    assert str(ver_range) == '<=4,1'
+
+    ver_range = VersionRange.from_list(['1'], is_right_closed=True)
+    assert str(ver_range) == '<=1'
+
+    ver_range = VersionRange.from_list(['1'], is_right_closed=True, is_left_closed=True)
+    assert str(ver_range) == '<=1,1'
+
+    with pytest.raises(ValueError):
+        VersionRange.from_list(['1', '2', '3', '4'])
+
+
+def test_classify_versions():
+    """Test classify_versions()."""
+    versions = classify_versions(['2', '3', '4', '1'], [VersionRange('<=2.1')])
+    assert len(versions) == 4
+    assert versions[0].version == '1'
+    assert versions[0].is_affected
+    assert versions[1].version == '2'
+    assert versions[1].is_affected
+    assert versions[2].version == '3'
+    assert not versions[2].is_affected
+    assert versions[3].version == '4'
+    assert not versions[3].is_affected
```

Comment:
Maybe this would be cleaner and more portable way
```python
expect_versions = ['1', '2', '3', '4']
expect_affected = [True, True, False, False]
for idx, v in enumerate(versions):
    assert v.version == expect_versions[idx]
    assert v.is_affected == expect_affected[idx]
```

Reply:
Yup, I agree. Thanks :wink: Definitely more readable.